### ISSUE TITLE
Consider moving Molecule Shapes TitledPanel into Sun

### ISSUE THREAD

**Issue by jonathanolson** [phetsims/sun#200]

**Comment by pixelzoom** [comment]
To facilitate discussion, here are a couple of examples from molecules-shapes:

![screenshot_09](https://cloud.githubusercontent.com/assets/3046552/10034068/e809f68e-614a-11e5-9f76-7816620bdc9a.png)

**Comment by pixelzoom** [comment]
Do we really need an entirely new type of panel? It might be better to take the Java approach: have different border types that can be combined, and apply them to a sun.Panel.  In Java, the "Bonding" panel above would be something like:

```js
var lineBorder = BorderFactory.createLineBorder( Color.white );
var titledBorder = BorderFactory.createTitledBorder( lineBorder, 'Bonding' );
titledBorder.setTitleJustification( TitledBorder.CENTER );
JPanel panel = new JPanel();
panel.setBorder( titledBorder );
```

Looking at the current implementation... The JSdoc (see below) describes the constraints.  I can imagine needing something other than 'centered' for the title, and that doesn't seem to difficult to generalize. The background color requirement affects the general usefulness of this panel, but I don't see an easy way around it.

```js
/**
 * A Panel, but with an added {Node} title centered along the location of the top border. Additionally, the border
 * behind the title is hidden.
 *
 * NOTE: TitledPanel requires that its background color is the same as the background color BEHIND the panel, since it
 * is intended that there is no line in-between the two near where the title is. We use a rectangle with this background
 * color to hide the border, so that it blends well into the content around it.
 *
 * @author Jonathan Olson <jonathan.olson@colorado.edu>
 */
```

**Comment by pixelzoom** [comment]
Above I said:
> Do we really need an entirely new type of panel? It might be better to take the Java approach:

Even if we don't take the Java approach of separating out borders... It might be better to decide what the desired look is for panel titles for all sims. If this is it, then great, and this seems like a fine solution, and TitledPanel is a fine name.  But if the look is specific to molecule-shapes and a few other sims where the background color constrain is tolerable, then choose a name for this panel that isn't so general.

**Comment by samreid** [comment]
Using a decorator pattern seems much more flexible than built-in support specific to borders:
```js
var contentNode = new Panel(/*controls here*/);
var titleNode = ...;
var titledPanel = new TitledPanel(titleNode,contentNode,{titledPanelOptions:here});
```

**Comment by jonathanolson** [comment]
I would be more than happy with more flexible patterns to add borders/titles/etc. to panels. I can keep TitledPanel specific to Molecule Shapes for now, and refactor to use any improved setup in the future.

**Comment by samreid** [comment]
I added the `developer-meeting` label since I expect this to be a common user interface element for future sims (though I may be wrong and am somewhat surprised that it is only used in one HTML5 sim at the moment).

**Comment by pixelzoom** [comment]
As I said in https://github.com/phetsims/sun/issues/200#issuecomment-142449797, I think that whether this panel is the desired look for titled panels is a *design* issue.  So before we spend time discussing in a developer meeting, should we pass on to some of the designers for input?

**Comment by samreid** [comment]
> But if the look is specific to molecule-shapes and a few other sims where the background color constraint is tolerable, 

We can easily create a kite shape that is "open" at the top and hence does not require the background color trick.

> should we pass on to some of the designers for input?

Can you please be more specific in what you recommend here?

**Comment by pixelzoom** [comment]
> > should we pass on to some of the designers for input?
> Can you please be more specific in what you recommend here?

Get input from PhET simulations designers to determine if this is the desired look for titled panels in PhET simulations, before spending development resources discussing/generalizing/implementing something that may not in fact be a "look" that is useful in general.

**Comment by samreid** [comment]
@ariel-phet can you chime in on https://github.com/phetsims/sun/issues/200#issuecomment-143805649 ?  There are screenshots above to show what this looks like.

**Comment by ariel-phet** [comment]
Yes, this is definitely a desired look. This is kind of like those "most interesting man in the world" commericials...we don't always have titled panels, but when we do, this seems to be a common look in certain sims (especially those with a black background). For instance States of Matter also use this same look (and it tends to save a bit of vertical space). 

So simple answer, yes, it would be a nice look to generalize, and has been used successfully in multiple design cases.

**Comment by samreid** [comment]
@pixelzoom would you recommend further discussion with designers before expanding on the development discussion?

**Comment by pixelzoom** [comment]
Design issues:

(1) Does this need to be usable in situations where the background behind the panel doesn't match the panel's background? (Imo yes, otherwise this isn't very generally useful.)

(2) If the answer to (1) is "yes", how should it look? Provide mockups to demonstrate.

(3) Is title alignment a desired option?  If so, are 'left', 'center', 'right' sufficient? Or is more precise positioning desired?

(4) Is this "titled border" look potentially applicable to things other than panels? (Ie, should it be factored out of TitledPanel and be usable elsewhere.)

**Comment by samreid** [comment]
@ariel-phet can you comment on (1)-(3) and possibly (4) above?

**Comment by ariel-phet** [comment]
1) Possibly, but that use case has not yet come up

2) That would be nice, but I don't think mockups would be useful. Cannot really be designed in a vacuum, and would need to wait until a sim instance occurs where such a panel is desire and it does not make sense to match the background. I can think of ways to do it, but until seeing it within a sim, we would just be taking a best guess and spinning our wheels. If I had to take a shot, I would probably continue the boarder stroke around the title, so that it basically had a rounded rectangle around it. But that might look clutzy depending on the circumstances.

3) That seems like a good generalization. Center would be a good option, but for right and left, people are likely going to want precise alignment to align with check boxes and such.

4) No pending use case comes to mind, but it is certainly within the realm of possibility. For instance, maybe someone would just be dying to put a label on a carousel in this style...ugh. We tend to avoid having words in sims when possible, but in the couple of use cases thus far (Molecule shapes and SOM) the labels serve good pedagogical value. So like I said, possible, but I see no use cases on the horizon.

**Comment by pixelzoom** [comment]
> 1) Possibly, but that use case has not yet come up

That use case exists in other sims that have not been ported yet. See for example and of the "Faraday's Electromagnetic Lab" variations.

> 2) That would be nice, but I don't think mockups would be useful. Cannot really be designed in a vacuum, and would need to wait until a sim instance occurs where such a panel is desire and it does not make sense to match the background. 

I don't understand this comment. How difficult is it to try this type of component on different background colors?  The current implementation does *not* support this, and the major cost of this UI component is in how the title appears.  A quick-&-dirty example would be to simply change the background color in Molecule Shapes and see if it looks acceptable or awful.

> 3) That seems like a good generalization. Center would be a good option, but for right and left, people are likely going to want precise alignment to align with check boxes and such.

Good to know about alignment with panel contents. There are lots of possible ways to implement alignment, would be good to not have to redo in the future.

> 4) No pending use case comes to mind, but it is certainly within the realm of possibility. F

Seems like it may be safe to implement within the context of TitledPanel, but keep in mind that we might want to pull out this functionality in the future.

**Comment by pixelzoom** [comment]
Why does this issue keep getting assigned to me? @jonathanolson owns this.

**Comment by ariel-phet** [comment]
for 1) and 2) - very possible that the panels will go to this style of having the same background color.

We can try something, but I think the case that exists above is what is most likely to be used. As seen in Molecule Shape, SOM, and likely will also occur in Gas Properties.

In this particular case does not make sense to me to support until we are working on sim where we actually deem in necessary, which may or may not come up.

**Comment by ariel-phet** [comment]
@pixelzoom btw, I often just reassign an issue back and forth when discussing, that does not mean it is officially assigned to you. In this case you and are were discussing, so after it got assigned to me, I was throwing back to your court.

**Comment by pixelzoom** [comment]
9/29/15 dev meeting:
• deferred. as soon as we need it in some other sim, we'll generalize it.
